strawberryfields/api/post_processing.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2019-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides functions for post-processing samples.
+"""
+import numpy as np
+
+
+def number_expectation_pnr(photon_number_samples, modes=None):
+    """The expectation value of the number operator from PNR samples.
+
+    **Example:**
+
+    .. code-block:: python
+
+        samples = np.array([[2, 0, None],
+                            [2, 2, None],
+                            [2, 0, None],
+                            [0, 0, None]])
+
+    Getting the expectation value for these PNR samples:
+
+    >>> number_expectation_pnr(samples)
+    1.0
+
+    Args:
+        photon_number_samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        float: the expectation value from the samples
+    """
+    product_across_modes = _checks_and_get_product(photon_number_samples, modes=modes)
+    expval = product_across_modes.mean()
+    return expval
+
+
+def number_variance_pnr(photon_number_samples, modes=None):
+    """The variance of the number operator from PNR samples.
+
+    **Example:**
+
+    .. code-block:: python
+
+        samples = np.array([[2, 0, None],
+                            [2, 2, None],
+                            [2, 0, None],
+                            [0, 0, None]])
+
+    Getting the variance for these PNR samples:
+
+    >>> number_expectation_pnr(samples)
```

Comment:
hey @antalszava : Maybe I am missing something, but how does it now which modes to consider? Is it just the ones that do not have `None`s?

Reply:
hi @nquesada, yes, if `modes` was not explicitly defined, then the idea would be considering all the modes where there are no `None` values (the measured modes).

This will actually be simplified by #398, as that change will remove the `None` values from the returned samples completely and only samples for the measured modes will be present. My original thinking was that handling `None` values could be manageable in this PR, but I also found that some parts started to become more complex. Let me know if you'd rather we waited with going forward for this until #398 was merged. It might also make the reviewing process more seamless.

Reply:
I think it might make more sense to merge in #398 first, especially if it makes a lot of the validation and the checks here redundant!

Reply:
Sure! In the meantime, potential things I'd be curious to hear thoughts on (if there are any):
* function naming (e.g. having `number_expectation_pnr` as user facing but also having a `number_expectation` as a shorthand)
* import structure (adding the name `number_expectation_pnr` to `api/__init__.py`)
* file structure (having a `post_processing.py` in the `api` folder)

Reply:
A couple of thoughts:

* For the moment, if both functions have no differences in logic, we should just have a single function.

* I mentioned this above, but I don't think the API module is the right place for sample post-processing. I think (for the moment) the best thing is to turn `sf.utils` into a package, and have `post_processing.py` in there.

Reply:
thanks @josh146! So they are the same function, just allowing further names through module attributes, e.g. at the end of `post_processing.py`:
```python
number_expectation = number_expectation_pnr
```

I've added this because during the preliminary discussion it was suggested that there is some sort of indication to the sample type --- this way we have both a shorthand and a longer version

(Went for your second point with #413)